### PR TITLE
gh-106529: Split FOR_ITER_{LIST,TUPLE} into uops

### PR DESCRIPTION
**Pull request by gvanrossum**
Same recipe as FOR_ITER_RANGE. I refactored the Tier 2 transformation code to handle similar cases. I had to fix a bug in the space reservation code, and decided to add some macros and an extra guard rail for that.

* Issue: gh-106529

### CODE REVIEW

**Comment by gvanrossum** [comment]
@iritkatriel Do you have time to look at the reservation code? The rest is on auto-pilot (I'm going to add FOR_ITER_TUPLE to the same PR.

File: Python/bytecodes.c [hunk 1]
```diff
@@ -2405,52 +2406,102 @@ dummy_func(
...
-            /* Jump forward oparg, then skip following END_FOR instruction */
-            JUMPBY(oparg + 1);
-            DISPATCH();
-        end_for_iter_list:
-            // Common case: no jump, leave it to the code generator
         }
 
-        inst(FOR_ITER_TUPLE, (unused/1, iter -- iter, next)) {
+        // Only used by Tier 2
+        op(_ITER_EXHAUSTED_LIST, (iter -- iter, exhausted)) {
```
Comment (iritkatriel) [inline Python/bytecodes.c]: maybe `_ITER_IS_EXHAUSTED_LIST`?

File: Python/bytecodes.c [hunk 2]
```diff
@@ -2405,52 +2406,102 @@ dummy_func(
...
+            _PyListIterObject *it = (_PyListIterObject *)iter;
+            assert(Py_TYPE(iter) == &PyListIter_Type);
+            PyListObject *seq = it->it_seq;
+            assert(seq);
+            assert(it->it_index < PyList_GET_SIZE(seq));
+            next = Py_NewRef(PyList_GET_ITEM(seq, it->it_index++));
+        }
+
+        macro(FOR_ITER_LIST) =
+            unused/1 + _ITER_CHECK_LIST + _ITER_JUMP_LIST + _ITER_NEXT_LIST;
```
Comment (iritkatriel) [inline Python/bytecodes.c]: In https://github.com/python/cpython/pull/106677 Mark added a `_SKIP_CACHE` bytecode. Note also the multiline formatting of macros there.

File: Python/bytecodes.c [hunk 3]
```diff
@@ -2405,52 +2406,102 @@ dummy_func(
...
-            SKIP_OVER(INLINE_CACHE_ENTRIES_FOR_ITER);
-            /* Jump forward oparg, then skip following END_FOR instruction */
-            JUMPBY(oparg + 1);
-            DISPATCH();
-        end_for_iter_tuple:
-            // Common case: no jump, leave it to the code generator
         }
 
+        // Only used by Tier 2
+        op(_ITER_EXHAUSTED_TUPLE, (iter -- iter, exhausted)) {
```
Comment (iritkatriel) [inline Python/bytecodes.c]: `_ITER_IS_EXHAUSTED_TUPLE`?

File: Python/optimizer.c [hunk 4]
```diff
@@ -474,26 +494,45 @@ translate_bytecode_to_trace(
...
+                        exhausted_op = _ITER_EXHAUSTED_TUPLE;
+                        next_op = _ITER_NEXT_TUPLE;
+                        break;
+                    case FOR_ITER_RANGE:
+                        check_op = _ITER_CHECK_RANGE;
+                        exhausted_op = _ITER_EXHAUSTED_RANGE;
+                        next_op = _ITER_NEXT_RANGE;
+                        break;
+                    default:
+                        assert(0);
```
Comment (iritkatriel) [inline Python/optimizer.c]:
```suggestion
                        Py_UNREACHABLE();
```

File: Python/optimizer.c [hunk 5]
```diff
@@ -378,13 +378,17 @@ translate_bytecode_to_trace(
     _Py_CODEUNIT *initial_instr = instr;
     int trace_length = 0;
     int max_length = buffer_size;
+    int reserved = 0;
```
Comment (iritkatriel) [inline Python/optimizer.c:381]: The point of reserve is to make sure that you won't error out in the middle of an opcode's translation to uops?

Comment (gvanrossum) [reply Python/optimizer.c:381]: Correct. I'd rather reserve space ahead than having to bail out in the middle and undo some work already done (the undoing feels more brittle).

File: Python/bytecodes.c (hunk 2 above)
Comment (gvanrossum) [reply Python/bytecodes.c]: Hm, I wonder if @markshannon forgot that we have the `unused/1` syntax in macros? The code generator handles this correctly in both tiers, no new opcodes are needed.

Happy to change the formatting (we should then also change the formatting in the docs, which serve as examples, and elsewhere in bytecodes.c, whenever it doesn't fit on one line).

Comment (markshannon) [reply Python/bytecodes.c]: I had forgotten about (if I was ever aware of) `unused/1`.

`_SKIP_CACHE` may a bit more informative, although `unused/1 // Skip counter` is pretty clear too.

File: Python/bytecodes.c (hunk 3 above)
Comment (markshannon) [reply Python/bytecodes.c]: Or start with the verb "_IS" to make it clearer that it is a predicate?

File: Python/bytecodes.c (hunk 2 above)
Comment (gvanrossum) [reply Python/bytecodes.c]: > I had forgotten about (if I was ever aware of) `unused/1`.

IIRC it was in your first draft for the DSL. :-)

We can get rid of `_SKIP_CACHE` later.

**Comment by gvanrossum** [comment]
Assuming the tests pass this time I'll just merge.

**Pull request merged**